src/lib/rules/html-checker/html-checker.ts

```diff
@@ -0,0 +1,107 @@
+// Requirements
+// ------------------------------------------------------------------------------
+
+import { debug as d } from '../../utils/debug';
+import { IRule, IRuleBuilder } from '../../types'; // eslint-disable-line no-unused-vars
+import { RuleContext } from '../../rule-context'; // eslint-disable-line no-unused-vars
+
+import { ITargetFetchEnd, IScanEnd, IProblemLocation, Severity } from '../../types'; // eslint-disable-line no-unused-vars
+
+const debug = d(__filename);
+
+// ------------------------------------------------------------------------------
+// Public
+// ------------------------------------------------------------------------------
+
+const rule: IRuleBuilder = {
+    create(context: RuleContext): IRule {
+        /** The promise that represents the scan by html checker. */
+        let htmlCheckerPromise: Promise<any>;
+        /** The options to pass to the html checker. */
+        const scanOptions = {
+            data: '',
+            format: 'json'
+        };
+
+        type Error = { // eslint-disable-line no-unused-vars
+            extract: string, // code snippet
+            firstColumn: number,
+            lastLine: number,
+            hiliteStart: number,
+            message: string,
+            subType: string
+        };
+
+        const start = (data: ITargetFetchEnd) => {
+            const { response } = data;
+            const htmlChecker = require('html-validator');
+
+            scanOptions.data = response.body.content;
+            htmlCheckerPromise = htmlChecker(scanOptions);
+        };
+
+        const end = async (data: IScanEnd) => {
+            const { resource } = data;
+            let result;
+
+            if (!htmlCheckerPromise) {
+                return;
+            }
+
+            debug(`Waiting for Html Checker results for ${resource}`);
+
+            try {
+                result = await htmlCheckerPromise;
+            } catch (e) {
+                debug(`Error getting html checker result for ${resource}.`);
+                await context.report(resource, null, `Couldn't get results from Html Checker for ${resource}.`);
+
+                return;
+            }
+
+            debug(`Received Html Checker results for ${resource}`);
+
+            const reportPromises = result.messages.map((error: Error): Promise<void> => {
+                const position: IProblemLocation = {
+                    column: error.firstColumn,
+                    elementColumn: error.hiliteStart + 1,
+                    elementLine: 1, // We will pass in the single-line code snippet generated from the html checker, so the elementLine is always 1
+                    line: error.lastLine
+                };
+
+                return context.report(resource, null, error.message, null, position, Severity[error.subType], error.extract);
+            });
+
+            try {
+                await Promise.all(reportPromises);
```

Comment:
Can we add support for async iterator? They should be working on typescript 2.3 (check the release notes, there might be some hack to do to make it work everywhere).

Also, I'd refactor the reporting and problem location code to another method and then call it in the iteration.